File: implementations/typescript/packages/api-server/src/operations.ts
// SPDX-License-Identifier: Apache-2.0
import { randomUUID } from "node:crypto";
import { canonicalJson, validateJson } from "@psp-cdl/core";
import { WorkflowStore } from "./persistence.js";
import { workflowError, type WorkflowAuthority } from "./workflow.js";
import { ServiceError, identifier, type ServiceHost, type Principal, type OperationSnapshot } from "./service.js";

export interface SessionOperationHost extends WorkflowAuthority {
  authenticate:ServiceHost["authenticate"];
  now:ServiceHost["now"];
  snapshot(principal:Principal, session:Record<string,unknown>, operationId:string):Pick<OperationSnapshot,"verification"|"resources"|"expires">|Promise<Pick<OperationSnapshot,"verification"|"resources"|"expires">>;
}
interface Binding { tenantId:string; subjectId:string; sessionId:string; version:unknown; nodeId:unknown; nodeVersion:unknown; policyVersion:unknown; epoch:string; expires:number }
/** Host-only, bounded in-memory handles. Restart drops handles and requires fresh issuance. */
export class SessionOperations implements ServiceHost {
  private readonly entries=new Map<string,Binding>();
  constructor(private readonly store:WorkflowStore,private readonly host:SessionOperationHost,private readonly capacity=1024) {
    if(!Number.isSafeInteger(capacity)||capacity<1) throw new ServiceError("INVALID_CONFIGURATION",500);
  }
  authenticate(token:string) { return this.host.authenticate(token); }
  now() { const n=this.host.now(); if(!Number.isSafeInteger(n)||n<0) throw new ServiceError("INTERNAL_ERROR",500); return n; }
  private async session(principal:Principal,id:string) {
    try { return await this.store.execute({tenantId:principal.tenantId,subjectId:principal.subjectId},{action:"getSession",sessionId:id},c=>this.host.authorize(structuredClone(principal),c)); }
    catch(e) { throw workflowError(e); }
  }
  private async binding(principal:Principal,session:Record<string,unknown>,expires:number):Promise<Binding> {
    const policy=await this.host.policyVersion(structuredClone(principal));
    if(!identifier(policy)) throw new ServiceError("INTERNAL_ERROR",500);
    if(session.status!=="running"||session.policyVersion!==policy) throw new ServiceError("STALE_OPERATION",409);
    return {tenantId:principal.tenantId,subjectId:principal.subjectId,sessionId:session.sessionId as string,version:session.version,nodeId:session.nodeId,nodeVersion:session.nodeVersion,policyVersion:policy,epoch:this.store.epoch,expires};
  }
  async issue(principal:Principal,sessionId:string,expires:number):Promise<string> {
    const now=this.now();
    if(!Number.isSafeInteger(expires)||expires<=now) throw new ServiceError("STALE_OPERATION",409);
    const session=await this.session(principal,sessionId);
    if(expires>(session.expiresAt as number)) throw new ServiceError("INVALID_REQUEST",400);
    const entry=await this.binding(principal,session,expires);
    for(const [id,b] of this.entries) if(this.now()>=b.expires) this.entries.delete(id);
    if(this.entries.size>=this.capacity) throw new ServiceError("OPERATION_CAPACITY",503);
    const id=randomUUID();this.entries.set(id,entry);return id;
  }
  revoke(operationId:string):void { this.entries.delete(operationId); }
  async resolve(principal:Principal,id:string):Promise<OperationSnapshot|null> {
    const entry=this.entries.get(id);
    if(!entry||entry.tenantId!==principal.tenantId||entry.subjectId!==principal.subjectId) return null;
    const live=async()=>{
      if(this.entries.get(id)!==entry||this.now()>=entry.expires) throw new ServiceError("STALE_OPERATION",409);
      const session=await this.session(principal,entry.sessionId);
      if(canonicalJson(await this.binding(principal,session,entry.expires))!==canonicalJson(entry)) throw new ServiceError("STALE_OPERATION",409);
      return session;
    };
    const session=await live();
    // Resolve keys, evidence and capabilities anew, including revocation and external policy changes.
    const details=structuredClone(await this.host.snapshot(structuredClone(principal),validateJson(session) as Record<string,unknown>,id));
    await live();
    if(!Number.isSafeInteger(details.expires)||details.expires<=this.now()) throw new ServiceError("STALE_OPERATION",409);
    const required={"tenant-id":entry.tenantId,"operation-id":id,"policy-version":String(entry.policyVersion),"session-id":entry.sessionId,"session-version":String(entry.version),"node-id":String(entry.nodeId),"node-version":String(entry.nodeVersion)};
    return {...details,tenantId:entry.tenantId,subjectId:entry.subjectId,operationId:id,policyVersion:String(entry.policyVersion),expires:Math.min(details.expires,entry.expires),verification:{...details.verification,context:{...details.verification.context,...required},allowedAttributes:[...new Set([...details.verification.allowedAttributes,...Object.keys(required)])]}};
  }
}

File: implementations/python/packages/api-server/src/psp_cdl_api_server/operations.py
# SPDX-License-Identifier: Apache-2.0
"""Host-only session-bound operation handles. Restart requires fresh issuance."""
from copy import deepcopy
import uuid
from psp_cdl_core import canonical_json, validate_json
from .persistence import integer
from .workflow import workflow_error
from .service import ServiceError, identifier


class SessionOperations:
    def __init__(self, store, host, capacity=1024):
        if not integer(capacity) or capacity < 1:
            raise ServiceError("INVALID_CONFIGURATION", 500)
        self.store, self.host, self.capacity = store, host, capacity
        self._entries = {}

    def authenticate(self, token):
        return self.host.authenticate(token)

    def now(self):
        now = self.host.now()
        if not integer(now):
            raise ServiceError("INTERNAL_ERROR", 500)
        return now

    def _session(self, principal, session_id):
        try:
            return self.store.execute({k: principal[k] for k in ("tenantId", "subjectId")}, {"action": "getSession", "sessionId": session_id}, lambda c: self.host.authorize(deepcopy(principal), c))
        except Exception as exc:
            raise workflow_error(exc) from None

    def _binding(self, principal, session, expires):
        policy = self.host.policy_version(deepcopy(principal))
        if not identifier(policy):
            raise ServiceError("INTERNAL_ERROR", 500)
        if session["status"] != "running" or session["policyVersion"] != policy:
            raise ServiceError("STALE_OPERATION", 409)
        return {"tenantId": principal["tenantId"], "subjectId": principal["subjectId"], **{k: session[k] for k in ("sessionId", "version", "nodeId", "nodeVersion")}, "policyVersion": policy, "epoch": self.store.epoch, "expires": expires}

    def issue(self, principal, session_id, expires):
        now = self.now()
        if not integer(expires) or expires <= now:
            raise ServiceError("STALE_OPERATION", 409)
        session = self._session(principal, session_id)
        if expires > session["expiresAt"]:
            raise ServiceError("INVALID_REQUEST", 400)
        entry = self._binding(principal, session, expires)
        self._entries = {k: v for k, v in self._entries.items() if self.now() < v["expires"]}
        if len(self._entries) >= self.capacity:
            raise ServiceError("OPERATION_CAPACITY", 503)
        operation_id = str(uuid.uuid4())
        self._entries[operation_id] = entry
        return operation_id

    def revoke(self, operation_id):
        self._entries.pop(operation_id, None)

    def resolve(self, principal, operation_id):
        entry = self._entries.get(operation_id)
        if entry is None or any(entry[k] != principal[k] for k in ("tenantId", "subjectId")):
            return None

        def live():
            if self._entries.get(operation_id) is not entry or self.now() >= entry["expires"]:
                raise ServiceError("STALE_OPERATION", 409)
            session = self._session(principal, entry["sessionId"])
            if canonical_json(self._binding(principal, session, entry["expires"])) != canonical_json(entry):
                raise ServiceError("STALE_OPERATION", 409)
            return session

        session = live()
        details = deepcopy(self.host.snapshot(deepcopy(principal), validate_json(session), operation_id))
        live()
        if not integer(details["expires"]) or details["expires"] <= self.now():
            raise ServiceError("STALE_OPERATION", 409)
        required = {"tenant-id": entry["tenantId"], "operation-id": operation_id, "policy-version": entry["policyVersion"], "session-id": entry["sessionId"], "session-version": str(int(entry["version"])), "node-id": entry["nodeId"], "node-version": entry["nodeVersion"]}
        verification = details["verification"]
        return {**details, "tenantId": entry["tenantId"], "subjectId": entry["subjectId"], "operationId": operation_id, "policyVersion": entry["policyVersion"], "expires": int(min(details["expires"], entry["expires"])), "verification": {**verification, "context": {**verification["context"], **required}, "allowedAttributes": list(dict.fromkeys([*verification["allowedAttributes"], *required]))}}
